### DISCUSSION TITLE
Peanut crop lifecycle

### DISCUSSION THREAD

**Discussion by ParameshwariS-tiger** [APSIMInitiative/ApsimX#9133] [Q&A] [answered]
Hi all,

I am using the Peanut module and need to input a harvest window similar to the sowing window. Is there a "Harvest using a variable rule" feature, similar to "Sow using a variable rule", that allows us to specify the harvest window (from date and to date)?

Alternatively, is it possible to set the peanut lifecycle to 150 days (from sowing to harvest)?

Any suggestions would be appreciated. Thank you!

**Comment by par456** [comment]
I think you want something like this. This script lets you set a harvest window and will harvest the peanut either when they are ready for harvesting within that window, or on the last day of that window if you set MustHarvest. You can add other requirements to this as well with further IF statements.

```C#
using System;
using Models.Core;
using Models.PMF;
using APSIM.Shared.Utilities;
namespace Models
{
    [Serializable]
    public class Script : Model
    {
        [Link] Clock Clock;
        [Link] Plant Peanut;
        
        [Description("Start of harvest window (dd-mmm)")]
        public string StartDate { get; set; }

        [Description("End of harvest window (dd-mmm)")]
        public string EndDate { get; set; }
        
        [Description("Must harvest by end")]
        public bool MustHarvest { get; set; }

        [EventSubscribe("DoManagement")]
        private void OnDoManagement(object sender, EventArgs e)
        {
            
            if (DateUtilities.WithinDates(StartDate, Clock.Today, EndDate) && Peanut.IsAlive)
            {
                if (Peanut.IsReadyForHarvesting || (MustHarvest && DateUtilities.CompareDates(EndDate, Clock.Today) == 0))
                {
                   Peanut.Harvest();
                   Peanut.EndCrop();    
                }
            }
        }
    }
}
```

**Reply by ParameshwariS-tiger** [reply]
@par456 Thank you, I tried the above method but the harvest date(sept 8) falls before the start date(sept 13) once I run the simulation. Below is the script I used:

using Models.Soils.Nutrients;
using System;
using System.Linq;
using Models.Core;
using Models.PMF;
using Models.Soils;
using Models.Soils.Nutrients;
using Models.Utilities;
using APSIM.Shared.Utilities;

namespace Models
{
    [Serializable]
    public class Script : Model
    {
        [Link] Clock Clock;
        [Link] Plant Peanut;
        
        [Description("Start of harvest window (dd-mmm)")]
        public string StartDate { get; set; }

        [Description("End of harvest window (dd-mmm)")]
        public string EndDate { get; set; }
        
        [Description("Must harvest by end")]
        public bool MustHarvest { get; set; }
        
        [EventSubscribe("DoManagement")]
        private void OnDoManagement(object sender, EventArgs e)
        {
            StartDate = "13-Sep";
            EndDate = "06-Dec";
            MustHarvest = true;
            
            if (DateUtilities.WithinDates(StartDate, Clock.Today, EndDate) && Peanut.IsAlive)
            {
                if (Peanut.IsReadyForHarvesting || (MustHarvest && DateUtilities.CompareDates(EndDate, Clock.Today) == 0))
                {
                   Peanut.Harvest();
                   Peanut.EndCrop();    
                }
            }
        
        }
        
    }
}

Is the script above correct? And also is it possible to set the peanut lifecycle to 150 days (from sowing to harvest)?

**Reply by hol353** [reply]
Script looks ok. Please attach a simulation that demonstrates the problem.

**Reply by ParameshwariS-tiger** [reply]
Hi, below is the zip folder which contains apsim file with peanut crop simulation for Bacon county, Georgia

[bacon(sow_harvest_window_2009).zip](https://github.com/user-attachments/files/16387825/bacon.sow_harvest_window_2009.zip)

**Reply by ParameshwariS-tiger** [reply]
@hol353 Can you help me with this? I have tried multiple times but the harvest date falls before the window that I specified in the script

**Reply by par456** [reply]
Hi just following this up, could you send through your met file for this simulation? I'm assuming your running in the northern hemisphere since it's planting in May, the problem may be in the met file for why it's not responding like you are expecting.

**Reply by ParameshwariS-tiger** [reply]
Yes right, I have downloaded weather data from APSIM for Bacon County, Georgia. Below is the .met file;
[bacon.zip](https://github.com/user-attachments/files/16756976/bacon.zip)

**Reply by par456** [reply]
Just ran your simulation and everything seems to be working. It sowed on the 17th of May, then harvested on the 13th of September, the first day of your harvest window. It in fact was ready to harvest on the 8th of September, and I can see that it then waits the 5 days to meet the conditions of your harvest window.

Is the problem you have here that you want to slow down the phenology of peanut so that it doesn't mature before that window?

**Reply by ParameshwariS-tiger** [reply]
How to check for harvest date? I was checking the phenology stage name "HarvestRipe" which still falls on September 8th (in DataStore model)

Generally, in USA harvesting period of peanut crop is from September to November and Sowing is from April to June. But APSIM's estimation of harvest date seems to be early, hence I was trying to set a harvest window

**Reply by par456** [reply, marked as answer]
If you want to see all the events that happen during a simulation, have a look at the summary file, it'll list every time the crop changes stages or an event happens.

To actually modify the length of phenology would involve modifying the actual peanut model, which would involve determining the differences between the current model and what you have observed. Moving the harvest date doesn't really do what you want here, since once it hits the harvest ripe phenology stage, it won't grow much more afterwards anyway. It's ready after 113 days, which is much shorter than the 150 you are looking for.

So to resolve this we might need one of the people that worked on the peanut model to give some advice.

**Comment by yashvirchauhan** [comment]
Which genotype is being used in your simulation that is being harvested in 117 days?

**Reply by ParameshwariS-tiger** [reply]
I am using Florunner cultivar
